compiler/rustc_const_eval/src/transform/validate.rs

```diff
@@ -109,14 +109,17 @@ impl<'tcx> MirPass<'tcx> for Validator {
                 // If this turns out not to be true, please let compiler-errors know.
                 // It is possible to support, but requires some changes to the layout
                 // computation code.
-                cfg_checker.fail(
-                    Location::START,
-                    format!(
-                        "Coroutine layout differs from by-move coroutine layout:\n\
-                        layout: {layout:#?}\n\
-                        by_move_layout: {by_move_layout:#?}",
-                    ),
-                );
+                // NOTE(@dingxiangfei2009): However, given that upvars are now part of the layout
```

Comment:
This is not correct, and will cause codegen problems.

Reply:
The comment says "If this turns out not to be true, please let compiler-errors know.", but you just commented it out 😓 Please make sure this does not land in this state.

Reply:
I will add the appropriate tests here.

The next thing to do is to check the the byref captures and make sure that they do match up modulo references, and the other parts of the layout are the same.

So far, by checking the codegen results, I think the outputs look right but I will be extra careful! :pray:

Reply:
What do you mean by "they do match up modulo references"?

The problem here is that when we call `tcx.coroutine_layout()` it is not going to be correct for by-move async closures that capture references.

Reply:
Yes, that is going to be a problem. I think I should change the coroutine_layout query to accept the closure kind, too. You can see that in other places in this patch, I have to check the correct kind type to determine if the by-move layout or the non-by-move layout shall be used. I find it error-prone so I am going to change thatq.

Reply:
I believe that's the correct change, however I've already started to make that change (starting with https://github.com/rust-lang/rust/pull/123021) so I'll have a PR out in a few minutes...

Reply:
Thank you for the PR! I have rebased onto the work and it looks much cleaner! :bow: